File: skills/adkit/scripts/ads_skill/lib/merge_test.py
from __future__ import annotations

from ads_skill.lib.merge import (
    ApiIdea,
    MIN_VOLUME,
    comparison_key,
    union_candidates,
)


def _idea(phrase: str, volume: int = 20_000, comp: str = "LOW",
          low: int | None = 1_000_000, high: int | None = 2_000_000) -> ApiIdea:
    return ApiIdea(phrase=phrase, volume=volume, competition=comp, low_micros=low, high_micros=high)


def test_comparison_key_collapses_case_and_whitespace() -> None:
    assert comparison_key("Buy Now") == comparison_key("buy  now")
    assert comparison_key("  Sell My CAR  ") == "sell my car"


def test_union_drops_zero_volume_api_phrases() -> None:
    result = union_candidates(llm=(), api=[_idea("dead phrase", volume=0)])
    assert result == ()


def test_union_drops_api_below_min_volume() -> None:
    result = union_candidates(llm=(), api=[_idea("low vol", volume=MIN_VOLUME - 1)])
    assert result == ()


def test_union_keeps_api_at_min_volume() -> None:
    result = union_candidates(llm=(), api=[_idea("ok vol", volume=MIN_VOLUME)])
    assert len(result) == 1
    assert result[0].volume == MIN_VOLUME


def test_union_drops_api_phrases_over_80_chars() -> None:
    long = "a" * 81
    result = union_candidates(llm=(), api=[_idea(long, volume=50_000)])
    assert result == ()


def test_union_attributes_api_metrics_to_matching_llm_phrase() -> None:
    result = union_candidates(
        llm=["Buy Now"],
        api=[_idea("buy  now", volume=36_000, comp="HIGH", low=8_000_000, high=14_000_000)],
    )
    assert len(result) == 1
    c = result[0]
    assert c.phrase == "Buy Now"  # LLM casing preserved
    assert c.source == "both"
    assert c.volume == 36_000
    assert c.competition == "HIGH"


def test_union_keeps_api_only_phrases_and_drops_bare_llm() -> None:
    result = union_candidates(
        llm=["coffee maker"],  # no API backing → dropped, not kept bare
        api=[_idea("espresso machine", volume=20_000)],
    )
    phrases = {c.phrase for c in result}
    assert phrases == {"espresso machine"}


def test_union_drops_bare_llm_with_no_api_match() -> None:
    result = union_candidates(llm=["niche phrase"], api=())
    assert result == ()

File: skills/adkit/scripts/src/lib/merge.test.ts
import { describe, expect, it } from "vitest";
import { type ApiIdea, MIN_VOLUME, comparisonKey, unionCandidates } from "./merge.js";

function idea(
  phrase: string,
  volume = 20_000,
  comp = "LOW",
  low: number | null = 1_000_000,
  high: number | null = 2_000_000,
  conceptGroup: string | null = null,
): ApiIdea {
  return { phrase, volume, competition: comp, lowMicros: low, highMicros: high, conceptGroup };
}

describe("comparisonKey", () => {
  it("collapses case and whitespace", () => {
    expect(comparisonKey("Buy Now")).toBe(comparisonKey("buy  now"));
    expect(comparisonKey("  Sell My CAR  ")).toBe("sell my car");
  });
});

describe("unionCandidates", () => {
  it("drops zero-volume api phrases", () => {
    const result = unionCandidates([], [idea("dead phrase", 0)]);
    expect(result).toEqual([]);
  });

  it("drops api below min volume", () => {
    const result = unionCandidates([], [idea("low vol", MIN_VOLUME - 1)]);
    expect(result).toEqual([]);
  });

  it("keeps api at min volume", () => {
    const result = unionCandidates([], [idea("ok vol", MIN_VOLUME)]);
    expect(result).toHaveLength(1);
    expect(result[0].volume).toBe(MIN_VOLUME);
  });

  it("drops api phrases over 80 chars", () => {
    const long = "a".repeat(81);
    const result = unionCandidates([], [idea(long, 50_000)]);
    expect(result).toEqual([]);
  });

  it("attributes api metrics to matching llm phrase", () => {
    const result = unionCandidates(
      ["Buy Now"],
      [idea("buy  now", 36_000, "HIGH", 8_000_000, 14_000_000)],
    );
    expect(result).toHaveLength(1);
    const c = result[0];
    expect(c.phrase).toBe("Buy Now"); // LLM casing preserved
    expect(c.source).toBe("both");
    expect(c.volume).toBe(36_000);
    expect(c.competition).toBe("HIGH");
  });

  it("carries the api concept group through to matched and api-only candidates", () => {
    const both = unionCandidates(
      ["buy now"],
      [idea("buy now", 20_000, "LOW", 1_000_000, 2_000_000, "Purchase Intent")],
    );
    expect(both[0].conceptGroup).toBe("Purchase Intent");
    const apiOnly = unionCandidates(
      [],
      [idea("espresso machine", 20_000, "LOW", 1_000_000, 2_000_000, "Coffee Makers")],
    );
    expect(apiOnly[0].conceptGroup).toBe("Coffee Makers");
  });

  it("keeps api-only phrases and drops bare llm", () => {
    const result = unionCandidates(
      ["coffee maker"], // no API backing -> dropped, not kept bare
      [idea("espresso machine", 20_000)],
    );
    const phrases = new Set(result.map((c) => c.phrase));
    expect(phrases).toEqual(new Set(["espresso machine"]));
  });

  it("drops bare llm with no api match", () => {
    const result = unionCandidates(["niche phrase"], []);
    expect(result).toEqual([]);
  });
});
